axiom/nr_php_packages.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <math.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "nr_axiom.h"
+#include "nr_limits.h"
+#include "nr_php_packages.h"
+#include "util_memory.h"
+#include "util_minmax_heap.h"
+#include "util_vector.h"
+#include "util_hashmap.h"
+#include "util_hashmap_private.h"
+#include "util_strings.h"
+
+nr_php_package_t* nr_php_package_create(char* name, char* version) {
+  nr_php_package_t* p = NULL;
+
+  if (name == NULL || version == NULL) {
+    return NULL;
+  }
+
+  p = (nr_php_package_t*)nr_malloc(sizeof(nr_php_package_t));
+
+  p->package_name = nr_strdup(name);
+  p->package_version = nr_strdup(version);
+
+  return p;
+}
+
+void nr_php_package_destroy(nr_php_package_t* p) {
+  if (p != NULL) {
+    nr_free(p->package_name);
+    nr_free(p->package_version);
+    nr_free(p);
+  }
+}
+
+void nr_php_packages_add_package(nr_php_packages_t** h, nr_php_package_t* p) {
+  if (h == NULL || *h == NULL) {
+    *h = nr_hashmap_create((nr_hashmap_dtor_func_t)nr_php_package_destroy);
+  }
+
+  if (p == NULL || p->package_name == NULL || p->package_version == NULL) {
+    return;
+  }
+
+  nr_hashmap_set(*h, p->package_name, nr_strlen(p->package_name), p);
+}
+
+void nr_php_packages_destroy_package(nr_php_packages_t** h) {
+  nr_hashmap_destroy(h);
+}
+
+size_t nr_php_packages_count(nr_php_packages_t* h) {
+  return nr_hashmap_count(h);
+}
+
+size_t nr_hashmap_count_buckets(const nr_php_packages_t* hashmap) {
+  return (size_t)(1 << hashmap->log2_num_buckets);
+}
+
+int nr_php_packages_has_package(nr_php_packages_t* h,
+                                char* package_name,
+                                size_t package_len) {
+  return nr_hashmap_has(h, package_name, package_len);
+}
+
+char* nr_php_package_to_json(nr_php_package_t* package) {
+  char* json;
+  nrbuf_t* buf = nr_buffer_create(0, 0);
+
+  if (package == NULL || package->package_name == NULL
+      || package->package_version == NULL) {
+    return NULL;
+  }
+
```

Comment:
Defer creating the buffer only after ensuring it needs to be created. Otherwise line 79 causes mem leak:
```suggestion
  nrbuf_t* buf = NULL;

  if (package == NULL || package->package_name == NULL
      || package->package_version == NULL) {
    return NULL;
  }
  buf = nr_buffer_create(0, 0);
```